cache/blobs.go

```diff
@@ -224,3 +249,170 @@ func isTypeWindows(sr *immutableRef) bool {
+		if err = writeOverlayUpperdir(ctx, cw, upper, lower); err != nil {
+			return emptyDesc, errors.Wrap(err, "failed to write diff")
+		}
+	}
+
+	var commitopts []content.Opt
+	if labels != nil {
+		commitopts = append(commitopts, content.WithLabels(labels))
+	}
+	dgst := cw.Digest()
+	if err := cw.Commit(ctx, 0, dgst, commitopts...); err != nil {
+		if !errdefs.IsAlreadyExists(err) {
+			return emptyDesc, errors.Wrap(err, "failed to commit")
+		}
+	}
+	cinfo, err := sr.cm.ContentStore.Info(ctx, dgst)
+	if err != nil {
+		return emptyDesc, errors.Wrap(err, "failed to get info from content store")
+	}
+	if cinfo.Labels == nil {
+		cinfo.Labels = make(map[string]string)
+	}
+	// Set uncompressed label if digest already existed without label
+	if _, ok := cinfo.Labels[containerdUncompressed]; !ok {
+		cinfo.Labels[containerdUncompressed] = labels[containerdUncompressed]
+		if _, err := sr.cm.ContentStore.Update(ctx, cinfo, "labels."+containerdUncompressed); err != nil {
+			return emptyDesc, errors.Wrap(err, "error setting uncompressed label")
+		}
+	}
+
+	return ocispec.Descriptor{
+		MediaType: mediaType,
+		Size:      cinfo.Size,
+		Digest:    cinfo.Digest,
+	}, nil
+}
+
+func writeOverlayUpperdir(ctx context.Context, w io.Writer, upper string, lower []mount.Mount) error {
+	cw := archive.NewChangeWriter(w, upper)
+	changeFn := cw.HandleChange
+	err := mount.WithTempMount(ctx, lower, func(lowerRoot string) error {
+		return fs.DiffDirChanges(ctx, changeFn, lowerRoot, &fs.DiffDirOptions{
+			DiffDir:      upper,
+			DeleteChange: overlayDeleteChange,
+		})
+	})
+	if err != nil {
+		return err
+	}
+	return cw.Close()
+}
+
+func overlayDeleteChange(diffDir string, path string, base string, f os.FileInfo) (deleteFile string, skip bool, err error) {
+	// Check if this is a whiteout
+	if f.Mode()&os.ModeCharDevice != 0 {
+		if _, ok := f.Sys().(*syscall.Stat_t); ok {
+			maj, min, err := devices.DeviceInfo(f)
+			if err != nil {
+				return "", false, err
+			}
+			if maj == 0 && min == 0 {
+				// This file is deleted from base directory
+				if _, err := os.Stat(filepath.Join(base, path)); err != nil {
+					if !os.IsNotExist(err) {
+						return "", false, err
+					}
+					// This file doesn't exist even in the base dir. We don't need whiteout.
+					return "", true, nil
+				}
+				return path, false, nil
+			}
+		}
+	}
+
+	// Check if this is an opaque directory
+	if f.IsDir() {
+		opaque, err := sysx.LGetxattr(filepath.Join(diffDir, path), "trusted.overlay.opaque")
+		if err != nil {
+			if err != unix.ENODATA {
+				return "", false, errors.Wrapf(err, "failed to retrieve trusted.overlay.opaque attr")
+			}
+		} else if len(opaque) == 1 && opaque[0] == 'y' {
+			// Add this directory and an opaque whiteout file.
+			if _, err := os.Stat(filepath.Join(base, path)); err != nil {
+				if !os.IsNotExist(err) {
+					return "", false, err
+				}
+				// This file doesn't exist even in the base dir. We don't need whiteout.
+				return "", true, nil
```

Comment:
Why is skip set to true here? If this dir exists in upper but not base, shouldn't it be included as part of the diff (as a dir creation)?

Reply:
Thank you. Fixed this.